src/truss.cpp

```diff
@@ -0,0 +1,242 @@
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc.,  51 Franklin Street, Fifth Floor, Boston, MA
+  02110-1301 USA
+
+*/
+
+#include "truss.h"
+#include <iostream>
+#include <vector>
+#include <unordered_set>
+
+using namespace std;
+
+/**
+ * \function trussness
+ * \brief Find the "trussness" for every edge in the graph. This
+ * indicates the highest k-truss that the edge occurs in.
+ *
+ * \param graph The input graph.
+ * \param truss Pointer to initialized vector of truss values that will
+ * indicate the highest k-truss each edge occurs in. It will be resized as
+ * needed.
+ *
+ * \example examples/simple/igraph_truss.cpp
+ */
+int igraph_truss(const igraph_t* graph, igraph_vector_int_t* truss){
+  igraph_vector_int_t triangles, support;
+  igraph_vector_t eid, unpacked_triangles;
+
+  // List the triangles as vertex triplets.
+  igraph_vector_int_init(&triangles, 0);  // will be resized
+  igraph_list_triangles(graph, &triangles);
+  IGRAPH_FINALLY(igraph_vector_int_destroy, &triangles);
+
+  // Unpack the triangles from vertex list to edge list.
+  long int num_triangles = igraph_vector_int_size(&triangles);
+  igraph_vector_init(&unpacked_triangles, 2 * num_triangles);
+  unpack(&triangles, &unpacked_triangles);
+  IGRAPH_FINALLY(igraph_vector_int_destroy, &unpacked_triangles);
+  igraph_vector_int_destroy(&triangles);
+  IGRAPH_FINALLY_CLEAN(1);
+
+  // Get the edge ids of the unpacked triangles. Note: a given eid can occur
+  // multiple times in this list if it is in multiple triangles.
+  igraph_vector_init(&eid, num_triangles);
+  igraph_get_eids(graph, &eid, &unpacked_triangles, 0, 0, 1);
+  IGRAPH_FINALLY(igraph_vector_int_destroy, &eid);
+  igraph_vector_destroy(&unpacked_triangles);
+  IGRAPH_FINALLY_CLEAN(1);
+
+  // Compute the support of the edges.
+  igraph_vector_int_init(&support, igraph_ecount(graph));
+  compute_support(&eid, &support);
+  IGRAPH_FINALLY(igraph_vector_int_destroy, &support);
+  igraph_vector_destroy(&eid);
+  IGRAPH_FINALLY_CLEAN(1);
+
+  // Compute the truss of the edges.
+  igraph_vector_int_init(truss, igraph_ecount(graph));
+  trussness(graph, &support, truss);
+  igraph_vector_int_destroy(&support);
+  IGRAPH_FINALLY_CLEAN(1);
+
+  return 0;
+}
+
+// Unpack the triangles as a vector of vertices to be a vector of edges.
+// So, instead of the triangle specified as vertices [1, 2, 3], return the
+// edges as [1, 2, 1, 3, 2, 3] so that the support can be computed.
+void unpack(const igraph_vector_int_t *tri, igraph_vector_t *unpacked_tri) {
+  int j;
+  for (long int i = 0; i < igraph_vector_int_size(tri); i = i + 3){
+    j = i * 2;
```

Comment:
It would be more efficient to add `j = 0` in the initializer of the `for` loop and use `i += 3, j += 2` instead of performing a multiplication in every iteration. (The compiler might be able to optimize it away, but I'm not sure).

Reply:
I agree addition is faster. However, I don't see how `j += 2` accomplishes what is needed here because the first iteration is `i = 0, j = 0`, the second iteration needs to be `i = 3, j = 6`, the third iteration needs to be `i = 6, j = 12`, etc.

Reply:
Ah, sorry, I meant `j += 6` instead.